pulser/simulation/simulation.py

```diff
@@ -147,34 +245,77 @@ def prepare_dict() -> dict[str, np.ndarray]:
                     'phase': np.zeros(self._tot_duration)}
 
         def write_samples(slot: _TimeSlot,
-                          samples_dict: Mapping[str, np.ndarray]) -> None:
+                          samples_dict: Mapping[str, np.ndarray],
+                          *qid: Union[int, str],
+                          is_global_pulse: bool = False) -> None:
+            """Builds hamiltonian coefficients.
+
+            Taking into account, if necessary, noise effects, which are local
+            and depend on the qubit's id qid.
+            """
             _pulse = cast(Pulse, slot.type)
-            samples_dict['amp'][slot.ti:slot.tf] += _pulse.amplitude.samples
-            samples_dict['det'][slot.ti:slot.tf] += _pulse.detuning.samples
+            noise_det = 0.
+            noise_amp = 1.
+
+            for noise in self.config.noise:
+                if noise == 'doppler':
+                    noise_det += np.random.normal(
+                        0, self.config.doppler_sigma)
+                # Gaussian beam loss in amplitude for global pulses only
+                if noise == 'amplitude' and is_global_pulse:
+                    position = self._qdict[qid[0]]
+                    r = np.linalg.norm(position)
+                    w0 = self.config.laser_waist
+                    noise_amp = np.random.normal(
+                        1., 1.e-3) * np.exp(-(r/w0)**2)
+
+            samples_dict['amp'][slot.ti:slot.tf] = \
+                _pulse.amplitude.samples * noise_amp
+            samples_dict['det'][slot.ti:slot.tf] = \
+                _pulse.detuning.samples + noise_det
             samples_dict['phase'][slot.ti:slot.tf] = _pulse.phase
 
         for channel in self._seq.declared_channels:
             addr = self._seq.declared_channels[channel].addressing
             basis = self._seq.declared_channels[channel].basis
 
-            samples_dict = self.samples[addr][basis]
-
-            if addr == 'Global':
+            # Case of clean global simulations
+            if addr == 'Global' and not self.config.noise:
+                samples_dict = self.samples[addr][basis]
                 if not samples_dict:
                     samples_dict = prepare_dict()
                 for slot in self._seq._schedule[channel]:
                     if isinstance(slot.type, Pulse):
-                        write_samples(slot, samples_dict)
+                        write_samples(slot, samples_dict, is_global_pulse=True)
+                self.samples[addr][basis] = samples_dict
 
+            # Any noise : global becomes local for each qubit in the reg
+            # Since coefficients are modified locally by all noises
+            elif addr == 'Global':
```

Comment:
I think you can merge this block with the next one if you change some things, like:
```suggestion
            else:
                is_global = addr == 'Global'
                addr == "Local"
```
then you can use `is_global` on `write_samples`, the rest seems pretty much identical. You can also go back to having a single `self.samples[addr][basis] = samples_dict` line.